src/main/java/com/rednavis/vaadin/showcase/backend/service/password/PasswordUtils.java

```diff
@@ -0,0 +1,134 @@
+package com.rednavis.vaadin.showcase.backend.service.password;
+
+import java.security.NoSuchAlgorithmException;
+import java.security.SecureRandom;
+import java.security.spec.InvalidKeySpecException;
+import java.util.Base64;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+import lombok.experimental.UtilityClass;
+
+@UtilityClass
+class PasswordUtils {
+
+  private static final String PBKDF2_ALGORITHM = "PBKDF2WithHmacSHA1";
+  // These constants may be changed without breaking existing hashes.
+  private static final int SALT_BYTE_SIZE = 24;
+  private static final int HASH_BYTE_SIZE = 18;
+  private static final int PBKDF2_ITERATIONS = 64000;
+  // These constants define the encoding and may not be changed.
+  private static final int HASH_SECTIONS = 5;
+  private static final int HASH_ALGORITHM_INDEX = 0;
+  private static final int ITERATION_INDEX = 1;
+  private static final int HASH_SIZE_INDEX = 2;
+  private static final int SALT_INDEX = 3;
+  private static final int PBKDF2_INDEX = 4;
+
+  public static String createHash(String password) throws CannotPerformOperationException {
+    return createHash(password.toCharArray());
+  }
+
+  public static String createHash(char[] password) throws CannotPerformOperationException {
+    // Generate a random salt
+    SecureRandom random = new SecureRandom();
+    byte[] salt = new byte[SALT_BYTE_SIZE];
+    random.nextBytes(salt);
+
+    // Hash the password
+    byte[] hash = pbkdf2(password, salt, PBKDF2_ITERATIONS, HASH_BYTE_SIZE);
+    int hashSize = hash.length;
+
+    // format: algorithm:iterations:hashSize:salt:hash
+    return "sha1:" + PBKDF2_ITERATIONS + ":" + hashSize + ":" + toBase64(salt) + ":" + toBase64(hash);
+  }
+
+  public static boolean verifyPassword(String password, String correctHash) throws CannotPerformOperationException, InvalidHashException {
+    return verifyPassword(password.toCharArray(), correctHash);
+  }
+
+  public static boolean verifyPassword(char[] password, String correctHash) throws CannotPerformOperationException, InvalidHashException {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [The method 'verifyPassword(char, String)' has an NPath complexity of 256](https://app.codacy.com/gh/rednavis/vaadin-showcase/pullRequest?prid=4839326)